### PR TITLE
Add implementation-task context packs

### LINKED ISSUE #275: [P0] Add implementation-task context packs for issue-driven coding agents
## Priority
P0

## Type
Feature / Product direction

## Area
Context packs, retrieval, CLI, agent workflows

## Problem
Madar currently works best for explain-style prompts such as `How does X work?`.

But real AI coding-agent workflows are usually task-driven, not question-driven:

- implement issue #257
- add a new feature
- fix a bug
- write tests for this behavior
- refactor a module
- review a PR
- migrate an API

In these workflows, the agent does not only need an explanation. It needs the implementation surface: likely edit files, related tests, contracts, existing patterns, risk boundaries, and validation commands.

## Goal
Add a new task type for implementation work:

```bash
madar pack "Implement issue #257: Improve workflow-centric retrieval" --task implement
```

This should compile context for safe code changes, not just runtime explanation.

## Implementation-pack contents
An implementation pack should include:

- likely edit files
- related tests
- contracts/types/interfaces
- public API/CLI/MCP surfaces affected by the task
- existing similar implementation patterns
- runtime paths only when relevant
- risk boundaries / files to avoid changing
- validation commands
- acceptance criteria summary
- missing context warnings

## Expected behavior example
Given a task such as:

```text
Implement issue #257: Improve workflow-centric retrieval for business capability prompts.
```

Madar should retrieve likely implementation files such as:

- `src/runtime/retrieval-gate.ts`
- `src/runtime/retrieve.ts`
- `src/runtime/retrieve/slicing.ts`
- `src/infrastructure/context-pack-command.ts`
- `src/contracts/context-pack.ts` if output shape changes

And test files such as:

- `tests/unit/retrieval-gate.test.ts`
- `tests/unit/retrieve-slice-v1.test.ts`
- context-pack command tests if present
- benchmark/fixture tests if relevant

It should not only retrieve runtime report-generation nodes.

## Retrieval recipe differences
`--task explain` asks:

```text
What code explains this behavior?
```

`--task implement` should ask:

```text
What code must the agent understand before changing this safely?
```

So ranking must prioritize:

- edit surface
- tests
- contracts
- existing patterns
- public behavior
- validation
- regression risk

## Acceptance criteria
- `madar pack <prompt> --task implement` is supported.
- Implementation packs expose likely edit files separately from supporting context.
- Implementation packs expose likely test files separately from supporting context.
- Implementation packs include contracts/types/public surfaces where relevant.
- Implementation packs include validation command suggestions.
- Implementation packs include risk boundaries or cautions.
- Given issue-style text, Madar does not behave like a normal explain pack.
- Add unit/fixture tests proving an issue-driven prompt retrieves implementation surfaces rather than only runtime explanation nodes.

## Non-goals
- Do not require GitHub API integration in this issue.
- Do not require LLM-generated implementation plans.
- Do not remove or weaken existing `--task explain` behavior.
- Do not overfit only to issue #257 or GoValidate.

## Suggested labels
priority:P0, type:feature, area:context-pack, area:retrieval, area:dx

### PR DESCRIPTION
**Pull request by mohanagy**
## Summary
- add implementation-specific guidance sections to context packs for likely edits, tests, contracts/public surfaces, risks, validation, and cautions
- reuse the new guidance in both CLI and MCP stdio context_pack flows
- make relevant-files, feature-map, and risk-map task-aware so implement-mode packs use implement retrieval rather than explain defaults

## Testing
- npm run typecheck
- npm run build
- CI=1 npm run test:run

Closes #275

## Summary by CodeRabbit

* **New Features**
  * Implementation guidance for "implement" tasks: likely edit/test files, contracts/public-surface hints, validation commands, risk boundaries, acceptance criteria, cautions, and optional runtime context.

* **Behavior**
  * Guidance included in explain/context-pack and tool responses for implement tasks; task-kind and task-intent can filter retrieval and influence results.
  * Retrieval ordering made deterministic for tied scores.

* **Tests**
  * Unit tests added/updated to verify presence, shape, and ordering of implementation guidance.

[Review Change Stack](https://app.coderabbit.ai/change-stack/mohanagy/madar/pull/279?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

**Comment by mohanagy** [comment]
Investigating the review note against the current branch. The weakened assertion was only added because the ranked implement-pack fixture is drifting across environments; if the underlying ordering can be made deterministic again, I'll restore a strict `existing_patterns` check instead of keeping the permissive assertion.

File: tests/unit/context-pack-command.test.ts [hunk 1]
```diff
@@ -463,6 +502,72 @@ describe('context-pack-command', () => {
...
+    }))
+    expect(payload.implementation?.existing_patterns).toEqual(expect.any(Array))
+    if ((payload.implementation?.existing_patterns?.length ?? 0) > 0) {
+      expect(payload.implementation?.existing_patterns).toEqual(expect.arrayContaining([
+        expect.objectContaining({
+          kind: 'pattern',
+          source_file: expect.stringMatching(/^src\//),
+        }),
+      ]))
+    }
```
[the remark this answers is not in the document]
Comment (mohanagy) [reply tests/unit/context-pack-command.test.ts]: Fixed by making tied retrieval candidates deterministic instead of insertion-ordered, then restoring the strict `existing_patterns` assertion. Added a retrieve regression for tied lexical seed matches and kept the implement-pack test requiring a real pattern entry.
